doc/nightmare.md

```diff
@@ -1,5 +1,288 @@
+
+Examples of identifiers in a UI:
+
+     <Row id="section-patron" ...
+     <Field id="input-patron-identifier" ...
+     <Button id="clickable-find-patron" ...
+     <MultiColumnList id="list-patrons" ...
+     <Row id="section-item" ...
+     <Field id="input-item-barcode" ...
+     <Button id="clickable-add-item" ...
+     <MultiColumnList id="list-items-checked-out" ...
+     <Button id="clickable-done" ...
+
+and usage in a test script:
+
+```js
+it('should show error when scanning item before patron card', done => {
+  nightmare
+  .wait('#clickable-checkout-module')
+  .click('#clickable-checkout-module')
+  .wait('#input-item-barcode')
+  .insert('#input-item-barcode',"some-item-barcode")
+  .wait('#clickable-add-item')
+  .click('#clickable-add-item')
+  .wait('#section-patron div[class^="textfieldError"]')
+  .evaluate(function() {
+    var errorText = document.querySelector('#section-patron div[class^="textfieldError"]').innerText;
+    if (!errorText.startsWith("Please fill")) {
+      throw new Error("Error message not found for item entered before patron found");
+    }
+  })
+  .then(done)
+  .catch(done)
+})
+```
+
+## Writing robust tests
+
+Nightmare tests are fussy. A few guidelines can go a long way toward writing
+robust tests that succeed regardless of the environment they run in or the speed
+of the internet connection they have access to.
+
+### Avoid timers
+
+React's tendency to render a page multiple times can work against the way Nightmare
+tests operate because Nightmare may begin interacting with the page as soon as it
+renders the first time. Avoid the temptation to use timers, e.g. `.wait(1000)`,
+to wait for the page to "settle down"; at best this strategy is unreliable because
+it so environmentally dependent. Instead, `.wait()` for specific elements to appear
+before interacting with them. For example,
+
+```js
+.wait('#some-button-id')
+.click('#some-button-id')
+```
+
+### Do not use XPath
+
+Code in a Nightmare test either operates in the context of the test or the context
+of the browser. Within a function passed to `.wait()` or `.evaluate()`, you're in
+browser context and can operate on the DOM. In browser context, avoid using
+`document.evaluate` to execute an XPath query to retrieve some part of the DOM.
+Do not do this. It seems to function asynchronously and causes Nightmare to
+operate in some mysterious multi-threaded mode, causing tests results to become
+interleaved with one another. Yes, really.
+
+
+### Extract text from the DOM using .evaluate()
+
+`.evaluate(fx, av1, ... avn)` accepts a function and a list of argument to pass
+to the function, and returns a Promise. You can use this feature to extract text
+from the DOM by passing a function that parses the DOM and returns a value. This
+example from platform-core's loan-renewal test gets the index of an anchor on
+the page, e.g. "<a href='some-link'>some text</a>", given "some text" and then
```

Comment:
You may need to escape the `<a>` tag here.  It generates as an actual link to nowhere when viewing the rendered markdown.